ci/github-script/prepare: move more steps from workflow

wolfgangwalther opened the pull request:
Follow up to #434541. This allows testing these other steps locally as well.

Theoretically this would also allow to use the PR's `prepare` step in a different file easily. This was originally my idea, when thinking about splitting `pull_request` and `pull_request_target` triggers into separate files. I now found a solution without repeating the whole file, though, so that's not strictly needed anymore.

Still an improvement for development, I think. Also, there's another bigger change for `prepare` coming in, now that code owners is merged with the main PR workflow: Branch verification, aka does the target branch of the PR match the commits on the branch? Taking this into account, iteration on the whole prepare step is much easier locally as "one thing".

## Things done

- [x] Tested locally.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
actions/checkout: rename from actions/get-merge-commit

This moves the logic to "check whether the PR can be merged and
determine the relevant commits" into the PR / prepare job directly -
since that's the only place where it is supposed to be used.

Because of the if condition in get-merge-commit, this logic was run
multiple times, for example in the lint workflow, where only one of
targetSha or mergedSha was provided as input. However, this input was
thrown away directly. This might not be a big problem, because this was
not expensive, so far. But with the next commit, this will become more
so.

This also separates the logic a bit cleaner - `prepare` figures out all
the parameters for the whole PR workflow, while `checkout` handles the
consistency around these checkouts.

wolfgangwalther committed:
actions/checkout: rename inconsistent pinned-from input

wolfgangwalther committed:
ci/github-script/prepare: load systems

wolfgangwalther committed:
ci/github-script/prepare: classify branches

wolfgangwalther committed:
ci/github-script/prepare: determine changed files

MattSturgeon commented on ci/github-script/prepare.js:
```diff
@@ -49,6 +49,9 @@ module.exports = async ({ github, context, core }) => {
     core.setOutput('mergedSha', mergedSha)
     core.setOutput('targetSha', targetSha)
+
+    core.setOutput('systems', require('../supportedSystems.json'))
```
What happens when `ci/github-script` is sparse checked-out without `ci/supportedSystems.json`?

philiptaron replied:
Thank you!

MattSturgeon commented on ci/github-script/prepare.js:
```diff
@@ -1,3 +1,5 @@
+const { classify } = require('../supportedBranches.js')
```
Same question re: sparse-checkout. I assume this would fail to import.

MattSturgeon reviewed (commented):
Seems good.

I have a couple questions about importing files that I believe aren't always present in CI due to sparse checkouts.

wolfgangwalther committed:
ci/github-script/prepare: fix logging of branch classification

Logging objects to stdout is not possible with `core.info`, so we
fallback to `console.log` instead. There's no functional difference for
these anyway.

wolfgangwalther committed:
workflows/pr.prepare: specify cone mode explicitly

This is the default, but who's supposed to know about that? Let's make
it explicit to be clear about the availability of files in `ci/`.

wolfgangwalther committed:
actions/checkout: remove unused input types

The two as-at inputs, were not boolean anymore, but GHA didn't complain.
Input types are actually not a thing for composite actions - they are
only documented for callable workflows.
